src/basicfuns.jl

```diff
@@ -246,6 +249,66 @@ function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     end
 end
 
+"""
+    logsumexp_onepass(X)
+
+Compute `log(sum(exp, X))` in a numerically stable way that avoids intermediate under- and
+overflow.
+
+In contrast to [`logsumexp`](@ref) the result is computed using a single pass over the data.
+`X` should be an iterator of real numbers.
+
+# References
+
+[Sebastian Nowozin: Streaming Log-sum-exp Computation.](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)
+"""
+function logsumexp_onepass(X)
+    # fallback for empty collections
+    isempty(X) && return log(sum(X))
+
+    # perform single pass over the data
+    xmax, r = _logsumexp_onepass(X, Base.IteratorEltype(X))
+
+    return xmax + log(r)
+end
+
+# with initial element: required by CUDA
+function _logsumexp_onepass(X, ::Base.HasEltype)
+    # compute initial element
+    FT = float(eltype(X))
+    init = (FT(-Inf), zero(FT))
+    r_one = one(FT)
+
+    # perform single pass over the data
+    return mapreduce(_logsumexp_onepass_op, X; init=init) do x
+        return float(x), r_one
+    end
+end
+
+# without initial element
+function _logsumexp_onepass(X, ::Base.EltypeUnknown)
+    return mapreduce(_logsumexp_onepass_op, X) do x
+        _x = float(x)
+        return _x, one(_x)
+    end
+end
+
+function _logsumexp_onepass_op((xmax1, r1)::T, (xmax2, r2)::T) where {T<:Tuple}
+    if xmax1 < xmax2
+        xmax = xmax2
+        a = exp(xmax1 - xmax2)
+        r = r2 + ifelse(isone(r1), a, r1 * a)
```

Comment:
Maybe add a comment explaining that the `isone` check is there for performance? BTW, I'm surprised it makes things faster when using `ifelse`, since both branches are supposed to be evaluated in that case (I'd have used `?`).

Reply:
Have you noticed this one?

Reply:
Yes, I just wanted to postpone any changes to the algorithm until the general layout (regarding CUDA etc) was clear.